hail/python/hail/backend/service_backend.py

```diff
@@ -8,14 +10,46 @@
 
 from hailtop.config import get_deploy_config, get_user_config
 from hailtop.auth import service_auth_headers
-from hailtop.utils import (retry_response_returning_functions,
-                           external_requests_client_session)
+from hailtop.utils import async_to_blocking, retry_transient_errors
 from hail.ir.renderer import CSERenderer
 
 from .backend import Backend
 from ..hail_logging import PythonOnlyLogger
 
 
+class ServiceSocket:
+    def __init__(self, *, deploy_config=None):
+        if not deploy_config:
+            deploy_config = get_deploy_config()
+        self.url = deploy_config.base_url('query')
+        self.session = aiohttp.ClientSession(headers=service_auth_headers(deploy_config, 'query'))
+
+    def close(self):
+        async_to_blocking(self.session.close())
+        self.session = None
+
+    def handle_response(self, resp):
+        if resp.type == aiohttp.WSMsgType.CLOSED:
+            raise aiohttp.ServerDisconnectedError('Socket was closed by server. (code={resp.data})')
+        if resp.type == aiohttp.WSMsgType.ERROR:
+            raise FatalError(f'Error raised while waiting for response from server: {resp}.')
+        assert resp.type == aiohttp.WSMsgType.TEXT
+        return resp.data
+
+    async def async_request(self, endpoint, **data):
+        async with self.session.ws_connect(f'{self.url}/api/v1alpha/{endpoint}') as socket:
+            await socket.send_str(json.dumps(data))
+            resp = self.handle_response(await socket.receive())
+            assert resp == endpoint
+            result = json.loads(self.handle_response(await socket.receive()))
+            if result['status'] != 200:
```

Comment:
see my note below, but I think this will error when `'status'` is missing from the struct (in the error case).